sizeOptimizer.cpp

```diff
@@ -36,26 +37,134 @@ SizeOptimizer::SizeOptimizer(CudaVersion::CudaVersion version,
 std::vector<const Transform *> *SizeOptimizer::optimize(size_t nBest,
                                                         int maxPercIncrease,
                                                         int maxMemMB,
+                                                        bool disallowRotation,
+                                                        bool disallowOptimization,
+                                                        int dimensionCount,
                                                         bool squareOnly,
                                                         bool crop) {
   std::vector<GeneralTransform> preoptimized;
   for (auto in : input) {
     std::vector<GeneralTransform> *tmp =
-        optimizeXYZ(in, nBest, maxPercIncrease, squareOnly, crop);
+        optimizeXYZ(in, nBest + 20, maxPercIncrease, disallowRotation,
+                    disallowOptimization, dimensionCount, squareOnly, crop);
     preoptimized.insert(preoptimized.end(), tmp->begin(), tmp->end());
     delete tmp;
   }
   return optimizeN(&preoptimized, maxMemMB, nBest);
 }
 
+void SizeOptimizer::swapSizes(GeneralTransform &in) {
+  in.swapped2D = true;
+  in.originalX = in.X;
+  in.originalY = in.Y;
+  in.X = in.originalY;
+  in.Y = in.originalX;
+}
+
+bool SizeOptimizer::swapSizes2D(GeneralTransform &in, const Polynom &x, const Polynom &y) {
```

Comment:
I'm not very happy about this method. 
Please add documentation to the header file.
Also, the if block starting at line 74 is very long and hard to read and reason about.
Consider refactoring it by using a single decision variable for each case, e.g.:
```
const bool case1 = (!divisibleBy2X) && (kernelCountX <= 1.5) && (in.X <= in.Y) && (!divisibleBy2Y); // some comment explaining why are we checking these values, possibly referencing your master theses
...
if (case1 || case 17) { swapSizes(in);}
```